File: python/src/fds/protobuf/stach/extensions/v1/ColumnStachExtensionBuilder.py
import json

import google.protobuf.json_format
from fds.protobuf.stach.Package_pb2 import Package

from src.fds.protobuf.stach.extensions.v1.ColumnOrganizedStachExtension import ColumnOrganizedStachExtension


class ColumnStachExtensionBuilder:

    def set_package(self, pkg):
        """
        Sets the package to the stach extension builder. The input package can be instance of Package, or
        string or object representation of the Package.

        :param pkg: Package object in string or object form.
        :return: builder instance
        """
        if (type(pkg) is Package):
            self.package = pkg
        elif (type(pkg) is str):
            package = Package()
            google.protobuf.json_format.Parse(pkg, package, ignore_unknown_fields=True, descriptor_pool=None)
            self.package = package
        else:
            pkg = json.dumps(pkg)
            package = Package()
            google.protobuf.json_format.Parse(pkg, package, ignore_unknown_fields=True, descriptor_pool=None)
            self.package = package
        return self;

    def build(self):
        """
        Builds and returns the instance of ColumnOrganizedStachExtension class

        """
        return ColumnOrganizedStachExtension(self.package)

File: java/src/main/java/com/factset/protobuf/stach/extensions/ColumnStachExtensionBuilder.java
package com.factset.protobuf.stach.extensions;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.protobuf.GeneratedMessageV3;
import com.google.protobuf.InvalidProtocolBufferException;

public interface ColumnStachExtensionBuilder<T extends GeneratedMessageV3> {

    /**
     * Sets the Package object.
     * @param pkg package object of type Package
     * @return builder instance
     */
    ColumnStachExtensionBuilder setPackage(T pkg);

    /**
     * Sets the Package object by parsing raw object input.
     * @param pkg package object
     * @return builder instance
     * @throws JsonProcessingException
     * @throws InvalidProtocolBufferException
     */
    ColumnStachExtensionBuilder setPackage(Object pkg) throws JsonProcessingException, InvalidProtocolBufferException;

    /**
     * Set the Package object by parsing the input in string format.
     * @param pkgString string form of package object
     * @return builder instance
     * @throws InvalidProtocolBufferException
     */
    ColumnStachExtensionBuilder setPackage(String pkgString) throws InvalidProtocolBufferException;

    /**
     * Builds the stach extension and returns the instance.
     * @return instance of column organized stach extension class.
     */
    StachExtensions build();
}
